### PR TITLE
impl(pubsub): add `pubsub::PullAckHandler`

### PR DESCRIPTION
**Pull request by coryan**
This class will be returned from `Subscriber::Pull()`.

Part of the work for #7187

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10301)

### CODE REVIEW

File: google/cloud/pubsub/pull_ack_handler.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+   * unusable after 10 minutes). The returned future is satisfied when the retry
+   * loop completes.
+   *
+   * If exactly-once is not enabled, the request is handled on a best-effort
+   * basis.
+   *
+   * If the future is satisfied with an Okay `Status` **and** exactly-once
+   * delivery is enabled in the subscription, then the message will not be
+   * resent by Cloud Pub/Sub.  We remind the reader that Cloud Pub/Sub defaults
+   * to "at least once" delivery, that is, without exactly-once delivery, the
```
Comment (devbww) [inline google/cloud/pubsub/pull_ack_handler.h:72]: s/, that is/. That is/

File: google/cloud/pubsub/pull_ack_handler.h [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * This interface allows applications to acknowledge and reject messages that
+ * are provided by the Cloud Pub/Sub C++ client library to the application. Note
+ * that this class is move-able, to support applications that process messages
+ * asynchronously. However, this class is *not* copy-able, because messages can
+ * only be acknowledged or rejected exactly once.
+ *
+ * @par Thread Safety
+ *
+ * This class is *thread compatible*, only one thread should call non-const
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/pull_ack_handler.h]: Done.

File: google/cloud/pubsub/pull_ack_handler.h (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/pull_ack_handler.h]: Done.

File: google/cloud/pubsub/mocks/mock_pull_ack_handler.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * [mocked-link]: @ref
+ * google::cloud::pubsub_internal::PullAckHandler::Impl
+ *
+ * @see @ref subscriber-mock for an example using this class.
+ */
+class MockPullAckHandler : public pubsub::PullAckHandler::Impl {
+ public:
+  MOCK_METHOD(future<Status>, ack, (), (override));
+  MOCK_METHOD(future<Status>, nack, (), (override));
+  MOCK_METHOD(std::string, ack_id, ());
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/mocks/mock_pull_ack_handler.h]: Removed.  In passing, I think the other `*AckHandler` mocks have the same problem, but I cannot decide right now if removing these things would be breaking.  Fixed #10312

File: google/cloud/pubsub/internal/pull_ack_handler_impl.h [hunk 4]
```diff
@@ -41,7 +40,7 @@ class PullLeaseManager;
  *
  * This is an implementation detail, hidden from the application.
  */
-class PullAckHandlerImpl : public pubsub::ExactlyOnceAckHandler::Impl {
+class PullAckHandlerImpl : public pubsub::PullAckHandler::Impl {
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/pull_ack_handler_impl.h]: Renamed to `DefaultPullAckHandler`.

File: google/cloud/pubsub/internal/pull_ack_handler_impl.h [hunk 5]
```diff
@@ -41,7 +40,7 @@ class PullLeaseManager;
  *
  * This is an implementation detail, hidden from the application.
  */
-class PullAckHandlerImpl : public pubsub::ExactlyOnceAckHandler::Impl {
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/pull_ack_handler_impl.h:44]: Fixed, thanks.

**Review by devbww — APPROVED** [review]

**Pull request merged**